### PR TITLE
Recursive inlining via InlineTransform and associated fixes

### PR DESCRIPTION
**Pull request by mlange05**
This PR separates the various styles of source-inlining (elementals, internals, pragma-marked) into a dedicated `Transformation` class that can be configured independently. This key motivation for this is that this allows us to perform recursive multi-step inlining, where individual inlining steps are performed at the leaves first and then propagated up by traversing the `Scheduler` tree in reverse order.

As these steps where previously baked into the `SCCBaseTransformation` several of these steps have now been removed. Due to complex interplay in the ec-physics regression, things like associate resolution and dead-code-elimination have to now also be performed with the inlining steps. Until we have proper pipeline configuration, I have chosen to go the simple route and simply lump these in with the `InlineTransformation`, but would appreciate feedback on this. This refactoring also has show some (severe!) problems with the previous "elemental function" inliner, which are also fixed in this PR. 

Edit: As discussed offline, I've also added a first draft of a "housekeeping" transformation (`SanitiseTransformation`), for a easily configurable code harmonisation utility pass. I plan to extend this as we progress through the trafor cleanup and consolidation.

Edit: Turns out one of the small fixes actually has a some far-reaching consequences: The elemental inliner was originally not checking for true elemental functions, and it now does(!), but the problem is that our OFP frontend cannot deal with prefix keywords. As the elemental inliner is needed for C-transpilation though, this means a bunch of tests, including CLOUDSC, have to be disabled for OFP. I've had a go at retro-fitting this feature, but it's quite involed due to the broken scoping around type nodes in the OFP AST. I propose dropping OFP entirely now, but I'll track that as a separate issue.

### CODE REVIEW

File: loki/transform/transform_inline.py [hunk 1]
```diff
@@ -193,7 +285,7 @@ def inline_elemental_functions(routine):
     # Remove all module imports that have become obsolete now
     import_map = {}
     for im in FindNodes(Import).visit(routine.spec):
-        if all(hasattr(s, 'type') and s.type.dtype in removed_functions for s in im.symbols):
+        if im.symbols and all(hasattr(s, 'type') and s.type.dtype in removed_functions for s in im.symbols):
```
Comment (reuterbal) [inline loki/transform/transform_inline.py:288]: I think the `hasattr(s, 'type')` should no longer be necessary - that is a remnant of the times when we represented the symbol list in `Import` with names as strings.

File: transformations/transformations/single_column_coalesced.py [hunk 2]
```diff
@@ -308,27 +288,10 @@ def process_kernel(self, routine):
...
-
-        # Perform full source-inlining for pragma-marked subroutines
-        if self.inline_marked:
-            # When inlining we allow the horizontal dimension to alias, so that
-            # the de/re-vectorisation captures the shared vector dimension.
-            inline_marked_subroutines(routine, allowed_aliases=(self.horizontal.index,))
-
         # Associates at the highest level, so they don't interfere
         # with the sections we need to do for detecting subroutine calls
         resolve_associates(routine)
```
Comment (reuterbal) [inline transformations/transformations/single_column_coalesced.py:293]: What's the reason for not including `resolve_associates` in the `SanitiseTransformation`?

**Review by reuterbal — CHANGES_REQUESTED** [review]
This looks great and is much appreciated!

I left a few comments while browsing the changes, most of which are not strictly necessary to act upon other than some massaging of docstrings (marking this as Request changes for that). One thing to consider, though, would be to add a test for the new `SanitiseTransformation` to the Loki core's test base. It's currently only being invoked from the tests in the transformations sub-package.

File: loki/transform/transform_inline.py [hunk 3]
```diff
@@ -179,7 +268,10 @@ def inline_elemental_functions(routine):
 
     exprmap = {}
     for call in FindInlineCalls().visit(routine.body):
-        if call.procedure_type is not BasicType.DEFERRED:
+        if call.procedure_type is BasicType.DEFERRED:
```
[the remark this answers is not in the document]
Comment (mlange05) [reply loki/transform/transform_inline.py:271]: Yes, I'm planning an overhaul of the logging from transformations once the configuration changes are more complete (so we can select log-leve per trafo, etc.).

[the remark this answers is not in the document]
Comment (mlange05) [reply loki/transform/transform_inline.py:271]: Possibly worth capturing via an issue?

File: loki/transform/transform_inline.py [hunk 4]
```diff
@@ -32,10 +35,88 @@
...
+        Perform dead code elimination, where unreachable branches are
+        trimmed from the code; default@ True
+    allowed_aliases : tuple or list of str or :any:`Expression`, optional
+        List of variables that will not be renamed in the parent scope during
+        internal and pragma-driven inlining.
+    remove_imports : bool
+        Strip unused import symbols after pragma-inlining (optional, default: True)
+    external_only : bool, optional
+        Do not replace variables declared in the local scope when
+        inlining constants (default: True)
```
[the remark this answers is not in the document]
Comment (mlange05) [reply loki/transform/transform_inline.py]: Yes!

File: transformations/transformations/single_column_coalesced.py (hunk 2 above)
Comment (mlange05) [reply transformations/transformations/single_column_coalesced.py]: Oversight, now fixed (I hope).

**Comment by mlange05** [comment]
Ok, I rebased and pushed the fixes to the suggestions - many thanks @reuterbal . I'll have to rebase/re-run this with ec-phys regression, just to make sure, but will ping here once confirmed.

**Comment by mlange05** [comment]
Ok, rebased, CI happy and confirmed with EC-physics regression. I've also added test for the new trafo and addressed all comments, except the breakage one. I think this is ready for another look, @reuterbal

File: loki/transform/transform_inline.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/transform/transform_inline.py:272]: #220

File: transformations/transformations/single_column_coalesced.py (hunk 2 above)
Comment (reuterbal) [reply transformations/transformations/single_column_coalesced.py:293]: Sure, I only asked for the reason, and that's a good enough one ;-)

**Review by reuterbal — APPROVED** [review]
Many thanks, changes look good to me and everything has been addressed. GTG! :shipit:

**Pull request merged**